src/chainstate/stacks/boot/pox-3.clar

```diff
@@ -0,0 +1,1316 @@
+                (err ERR_INVALID_START_BURN_HEIGHT))
+
+      ;; must be called directly by the tx-sender or by an allowed contract-caller
+      (asserts! (check-caller-allowed)
+                (err ERR_STACKING_PERMISSION_DENIED))
+
+      ;; tx-sender principal must not be stacking
+      (asserts! (is-none (get-stacker-info tx-sender))
+        (err ERR_STACKING_ALREADY_STACKED))
+
+      ;; tx-sender must not be delegating
+      (asserts! (is-none (get-check-delegation tx-sender))
+        (err ERR_STACKING_ALREADY_DELEGATED))
+
+      ;; the Stacker must have sufficient unlocked funds
+      (asserts! (>= (stx-get-balance tx-sender) amount-ustx)
+        (err ERR_STACKING_INSUFFICIENT_FUNDS))
+
+      ;; ensure that stacking can be performed
+      (try! (can-stack-stx pox-addr amount-ustx first-reward-cycle lock-period))
+
+      ;; register the PoX address with the amount stacked
+      (let ((reward-set-indexes (try! (add-pox-addr-to-reward-cycles pox-addr first-reward-cycle lock-period amount-ustx tx-sender))))
+          ;; add stacker record
+         (map-set stacking-state
+           { stacker: tx-sender }
+           { pox-addr: pox-addr,
+             reward-set-indexes: reward-set-indexes,
+             first-reward-cycle: first-reward-cycle,
+             lock-period: lock-period,
+             delegated-to: none })
+
+          ;; return the lock-up information, so the node can actually carry out the lock.
+          (ok { stacker: tx-sender, lock-amount: amount-ustx, unlock-burn-height: (reward-cycle-to-burn-height (+ first-reward-cycle lock-period)) }))))
+
+(define-public (revoke-delegate-stx)
+  (begin
+    ;; must be called directly by the tx-sender or by an allowed contract-caller
+    (asserts! (check-caller-allowed)
+              (err ERR_STACKING_PERMISSION_DENIED))
+    (ok (map-delete delegation-state { stacker: tx-sender }))))
+
+;; Delegate to `delegate-to` the ability to stack from a given address.
+;;  This method _does not_ lock the funds, rather, it allows the delegate
+;;  to issue the stacking lock.
+;; The caller specifies:
+;;   * amount-ustx: the total amount of ustx the delegate may be allowed to lock
+;;   * until-burn-ht: an optional burn height at which this delegation expires
+;;   * pox-addr: an optional address to which any rewards *must* be sent
+(define-public (delegate-stx (amount-ustx uint)
+                             (delegate-to principal)
+                             (until-burn-ht (optional uint))
+                             (pox-addr (optional { version: (buff 1),
+                                                   hashbytes: (buff 32) })))
+    (begin
+      ;; must be called directly by the tx-sender or by an allowed contract-caller
+      (asserts! (check-caller-allowed)
+                (err ERR_STACKING_PERMISSION_DENIED))
+
+      ;; tx-sender principal must not be stacking
+      (asserts! (is-none (get-stacker-info tx-sender))
```

Comment:
Why that? This is a regression. In pox-2 it was possible to change the delegation amount while stacking.
https://github.com/stacks-network/stacks-blockchain/blob/d657bdde9bbe263050d2e7ffa4b822fdaefa2530/src/chainstate/stacks/boot/pox-2.clar#L651

Reply:
This avoids the problem that was present in auto-unlock.

Reply:
It takes a lot of power from delegate-stack-increase

Reply:
@obycode would it be enough to check that the pool does not change?

Reply:
I don't think so -- the issue is that `auto-unlock` relies on the invariant that if a user's solo stacking amount doesn't reach the participation threshold, then their account can be unlocked. But if they've also contributed to a pool, then their account would be unlocked _while still contributing to a reward_.

Reply:
Friedger and I were discussing this in Discord. What he would like is for users to be able to re-delegate to the same pool without a cooldown cycle. Essentially, this would be the equivalent of `stack-increase` and/or `stack-extend` but for delegation.

Reply:
But isn't that use case fulfilled with `delegate-stack-increase`?

Reply:
Also the current code is not convert by testing using devent-js

Reply:
All tests should pass if the assertion is removed. The important assertions should be in delegate-stack-* functions